### DISCUSSION TITLE
Git status not updating after :G commit with fugitive

### DISCUSSION THREAD

**Discussion by Theo-Steiner** [nvim-neo-tree/neo-tree.nvim#322] [Q&A] [answered]
Hey, thank you so much for writing this plugin.
The git status of a repo is updated pretty promptly when I for example stage a change with fugitive ``:Gwrite`` or restore with ``Gread``, but after committing with ``G commit``, the stale git info remains in the tree until I toggle it off and on once.
Is there a way to perhaps detect this as well? I tried enabling the use_libuv_file_watcher option or git_status_async but so far to no avail...

**Comment by nyngwang** [comment]
I have the same problem.

**Comment by cseickel** [comment]
The bad news is that there is not normally any way detect a git change that does not modify files on the file system. I tried adding a watcher on the git index but the git index but git is complicated...

The good news is that Fugitive actually fires a custom user event anytime it does anything that might change git status, further establishing why tpope is a God amongst plugin developers (thanks tpope!). You can hook into that event and automatically refresh Neotree like this:

```lua
vim.cmd[[
augroup FugitiveRefreshNeotree
  autocmd!
  autocmd User FugitiveChanged lua require("neo-tree.sources.manager").refresh("filesystem")
augroup END
]]
```

There is of course a pure lua equivalent now in 0.7, but I haven't started using them yet. 

I'll look into whether I can detect the presence of Fugitive and subscribe to this event automatically if Fugitive is installed.

**Comment by cseickel** [comment, marked as answer]
I actually just built this into Neo-tree, it is in `main` right now and will go out on the next release. It will now refresh on any Fugitive command that fires the `FugitiveChanged` event, which should be anything that can alter git status.

**Reply by Theo-Steiner** [reply]
You are a legend!

**Comment by bennypowers** [comment]
How about a periodic refresh?

I switched from Fugutive to lazygit in a scratch terminal (it's ok i guess). From my side, I'll look into refreshing neo-tree imperatively when that scratch terminal closes, shouldn't be too hard. But, this morning I noticed that a commit I made and pushed last night was still orange (marked as staged) in neo-tree's fs view. 

A once-every-5-minutes or once-every-hour refresh shouldn't be too heavy on resources, no?

**Reply by cseickel** [reply]
Maybe. I think I'll take another look at the options for an automatic git refresh. In the meantime, you could add this into your config to get a periodic refresh:

```lua
  if vim.g.neo_tree_timer then
    pcall(vim.g.neo_tree_timer.close, vim.g.neo_tree_timer)
  end
  -- Create a timer handle (implementation detail: uv_timer_t).
  local timer = vim.loop.new_timer()
  -- Waits 60 seconds, then repeats every 10 seconds until timer:close().
  timer:start(60 * 1000, 10 * 1000, function()
    require("neo-tree.events").fire_event("git_event")
  end)
  vim.g.neo_tree_timer = timer

```

**Reply by bennypowers** [reply]
And for those watching at home, here's how I finagled `lazygit` + `toggleterm` to refresh neo-tree when closing the lazygit terminal:

```lua
local Terminal = require 'toggleterm.terminal'.Terminal

---Scratch terminal for lazygit.
---Closes when lazygit exits cleanly.
---For good measure, refresh the neo-tree filesystem view either way
local lazygit = Terminal:new({
  cmd = 'lazygit',
  direction = 'float',
  hidden = true,
  on_exit = function (terminal, _, exit_code)
    if exit_code == 0 then
      terminal:close()
    end
    require 'neo-tree.sources.manager'.refresh 'filesystem'
  end,
})

local M = {}

---Launch lazygit in a scratch terminal,
function M.lazilygit()
  lazygit:toggle()
end

return M
```
